relm4-macros/src/util.rs

```diff
@@ -2,7 +2,53 @@ use proc_macro::TokenStream;
 use proc_macro2::Span as Span2;
 use syn::punctuated::Punctuated;
 
-use syn::{Ident, Path, PathArguments, PathSegment};
+use syn::spanned::Spanned;
+use syn::{Ident, ImplItem, ItemImpl, Path, PathArguments, PathSegment, Type, TypePath};
+
+pub(super) fn generate_widgets_type(
+    widgets_ty: Option<Type>,
+    component_impl: &mut ItemImpl,
+    errors: &mut Vec<syn::Error>,
+) -> Option<Type> {
+    // Use the widget type if used.
+    if let Some(ty) = widgets_ty {
+        Some(ty)
+    }
+    // Try to generate the type from the self type name.
+    else if let Type::Path(self_ty) = &*component_impl.self_ty {
+        let (path, impl_item) = self_ty_to_widgets_ty(self_ty);
+        component_impl.items.push(impl_item);
+        Some(path)
+    }
+    // Error: No Widget type found or generated.
+    else {
+        let msg = "No `Widgets` found. Please use `type Widgets = WidgetsName;` to name the widgets type.";
```

Comment:
This error message seems inaccurate, wouldn't it only be caused if `self_ty` isn't a normal path?

We should probably validate that earlier if we don't already, and emit an "invalid type for impl" error there. Then we could just unwrap here.

Reply:
Actually, the type of `Self` doesn't need to be a path: https://doc.rust-lang.org/reference/items/implementations.html#implementations

I've improved the error message to reflect the actual causes for the error. But apart from that, this is kind of a best effort strategy. As long as a path is used for `Self`, we can generate a `Widgets` type but for any other type of type (which is going to be extremely rare), we don't know what to do and leave the naming up to the user.